chips/apollo3/src/flash_ctrl.rs

```diff
@@ -0,0 +1,350 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Tock Contributors 2022.
+
+//! Flash Controller
+
+use core::cell::Cell;
+use core::ops::{Index, IndexMut};
+use kernel::deferred_call::{DeferredCall, DeferredCallClient};
+use kernel::hil;
+use kernel::utilities::cells::OptionalCell;
+use kernel::utilities::cells::TakeCell;
+use kernel::ErrorCode;
+
+pub const PAGE_SIZE: usize = 8 * 1024;
+pub const FLASH_INSTANCE_SIZE: usize = 512 * 1024;
+pub const FLASH_NUM_INSTANCES: usize = 2;
+pub const FLASH_PAGES_PER_INSTANCE: usize = FLASH_INSTANCE_SIZE / PAGE_SIZE;
+pub const FLASH_MAX_PAGES: usize = FLASH_NUM_INSTANCES * FLASH_PAGES_PER_INSTANCE;
+
+const FLASH_PROGRAM_KEY: u32 = 0x12344321;
+
+/// There are two flash instances, each is 512KiB.
+#[derive(PartialEq)]
+enum FlashInstance {
+    MAIN0 = 0,
+    MAIN1 = 1,
+}
+
+pub struct Apollo3Page(pub [u8; PAGE_SIZE]);
+
+impl Default for Apollo3Page {
+    fn default() -> Self {
+        Self([0; PAGE_SIZE])
+    }
+}
+
+impl Index<usize> for Apollo3Page {
+    type Output = u8;
+
+    fn index(&self, idx: usize) -> &u8 {
+        &self.0[idx]
+    }
+}
+
+impl IndexMut<usize> for Apollo3Page {
+    fn index_mut(&mut self, idx: usize) -> &mut u8 {
+        &mut self.0[idx]
+    }
+}
+
+impl AsMut<[u8]> for Apollo3Page {
+    fn as_mut(&mut self) -> &mut [u8] {
+        &mut self.0
+    }
+}
+
+/// The below documentation is based on the official HAL documentation.
+///
+/// This function will call the chip ROM code to perform the below operation.
+///
+/// Use this function to safely read a value from peripheral or memory locations.
+///
+/// @param addr - The location to be read.
+///
+/// @return The value read from the given address.
+unsafe fn flash_util_read_word(addr: *mut u32) -> u32 {
+    // Call `uint32_t flash_util_read_word(uint32_t *)` in the ROM code.
+    let flash_util_read_word: unsafe extern "C" fn(*mut u32) -> u32 =
+        unsafe { core::mem::transmute(0x08000075 as *const ()) };
+
+    flash_util_read_word(addr)
+}
```

Comment:
Why are these functions marked unsafe? Isn't the unsafe operation mapping the function?

I'm not sure these can/should be marked safe, but I think that would be ideal as it isn't clear to me what the caller can do to ensure the operation is safe.

Reply:
They can be used to read / write arbitrary flash memory, and thus arbitrary program code. As such, they are unsafe operations -- in much the same spirit that `mmap` is unsafe in Rust.

Reply:
Got it.

Seems like `read_word()` is fine (ie safe).

The others should just copy what you wrote into the comment.

Reply:
We don't know that though, so I do think it makes sense to keep them all unsafe

Reply:
> Seems like read_word() is fine (ie safe).

I think it should be unsafe as well.

This allows any capsule (with access to the method which, admittadly, it is private now) to read any flash data, which, in general, violates memory safety. It's only safe in the specific context that the `addr` points to a valid address that the caller is allowed to read, as  `u32`.